src/equal.c

```diff
@@ -1,33 +1,52 @@
 #include <math.h>
+#include "equal.h"
 #include "vctrs.h"
 
-static int lgl_equal_scalar(const int* x, const int* y, bool na_equal);
-static int int_equal_scalar(const int* x, const int* y, bool na_equal);
-static int dbl_equal_scalar(const double* x, const double* y, bool na_equal);
-static int raw_equal_scalar(const Rbyte* x, const Rbyte* y, bool na_equal);
-static int cpl_equal_scalar(const Rcomplex* x, const Rcomplex* y, bool na_equal);
-static int chr_equal_scalar(const SEXP* x, const SEXP* y, bool na_equal);
-static int list_equal_scalar(SEXP x, R_len_t i, SEXP y, R_len_t j, bool na_equal);
-static int df_equal_scalar(SEXP x, R_len_t i, SEXP y, R_len_t j, bool na_equal, int n_col);
-
 
 // If `x` is a data frame, it must have been recursively proxied
 // beforehand
 //
 // [[ include("vctrs.h") ]]
 int equal_scalar(SEXP x, R_len_t i, SEXP y, R_len_t j, bool na_equal) {
-  switch (TYPEOF(x)) {
-  case LGLSXP: return lgl_equal_scalar(LOGICAL(x) + i, LOGICAL(y) + j, na_equal);
-  case INTSXP: return int_equal_scalar(INTEGER(x) + i, INTEGER(y) + j, na_equal);
-  case REALSXP: return dbl_equal_scalar(REAL(x) + i, REAL(y) + j, na_equal);
-  case STRSXP: return chr_equal_scalar(STRING_PTR(x) + i, STRING_PTR(y) + j, na_equal);
-  case RAWSXP: return raw_equal_scalar(RAW(x) + i, RAW(y) + j, na_equal);
-  case CPLXSXP: return cpl_equal_scalar(COMPLEX(x) + i, COMPLEX(y) + j, na_equal);
+  enum vctrs_type proxy_type = vec_proxy_typeof(x);
+  const void* x_p = NULL;
+  const void* y_p = NULL;
+
+  switch (proxy_type) {
```

Comment:
I assume you've made sure all these conditions are covered by tests?

Reply:
We were missing coverage for lgl, cpl, and raw. Now fixed.